substrate/frame/support/procedural/src/construct_runtime/expand/hold_reason.rs

```diff
@@ -61,7 +71,14 @@ pub fn expand_outer_hold_reason(pallet_decls: &[Pallet], scrate: &TokenStream) -
 		}
 
 		impl #scrate::traits::VariantCount for RuntimeHoldReason {
-			const VARIANT_COUNT: u32 = #hold_reason_variants_count;
+			const VARIANT_COUNT: u32 = #( #hold_reason_variants_count )+*;
+		}
+
+		/// Implementation of `Get<u32>` which returns `RuntimeHoldReason::VARIANT_COUNT`.
+		impl #scrate::traits::Get<u32> for RuntimeHoldReason {
+			fn get() -> u32 {
+				<RuntimeHoldReason as #scrate::traits::VariantCount>::VARIANT_COUNT
+			}
 		}
```

Comment:
I wouldn't do this. 

Instead I would remove `MaxHolds` and `MaxFreezes` and require that `RuntimeHoldReason` and `RuntimeFreezeReason` implement `VariantCount`.

Reply:
@bkchr Yes, exactly what I had in mind when adding this `Get` implementation, but I just wanted to be cautious and avoid introducing a **breaking change** at the beginning :)

`RuntimeFreezeReason` or `RuntimeHoldReason` already implements `VariantCount` and `pallet_balances` defines them as:
```
type RuntimeHoldReason: Parameter + Member + MaxEncodedLen + Copy + VariantCount;
type RuntimeFreezeReason: VariantCount;
```

For `Holds`, I tried to replace `T::MaxHolds` with `VariantCount` with `ConstU32`:
```
pub type Holds<T: Config<I>, I: 'static = ()> = StorageMap<
        ...
        // BoundedVec<IdAmount<T::RuntimeHoldReason, T::Balance>, T::MaxHolds>,
        BoundedVec<IdAmount<T::RuntimeHoldReason, T::Balance>, ConstU32<{ <T::RuntimeHoldReason as VariantCount>::VARIANT_COUNT }>>,,
        ...
>;
```

But this does not work:
```
error: generic parameters may not be used in const operations
   --> substrate/frame/balances/src/lib.rs:483:70
    |
483 |         BoundedVec<IdAmount<T::RuntimeHoldReason, T::Balance>, ConstU32<{ <T::RuntimeHoldReason as VariantCount>::VARIANT_COUNT }>>,
    |                                                                            ^^^^^^^^^^^^^^^^^^^^ cannot perform const operation using `T`
    |
    = note: type parameters may not be used in const expressions
```

But with `VariantCountOf` adapter, it works:
```
/// Adapter for `Get<u32>` to access `VARIANT_COUNT` from `trait pub trait VariantCount {`.
pub struct VariantCountOf<T: VariantCount>(sp_std::marker::PhantomData<T>);
impl<T: VariantCount> Get<u32> for VariantCountOf<T> {
	fn get() -> u32 {
		T::VARIANT_COUNT
	}
}

...

pub type Holds<T: Config<I>, I: 'static = ()> = StorageMap<
        ...
        // BoundedVec<IdAmount<T::RuntimeHoldReason, T::Balance>, T::MaxHolds>,
        BoundedVec<IdAmount<T::RuntimeHoldReason, T::Balance>, VariantCountOf<T::RuntimeHoldReason>>,
        ...
>;
```

Please, check commit: https://github.com/paritytech/polkadot-sdk/pull/2657/commits/9bde28f53fdce648d9e13bbc46c4dd8ce6a352cc

--- 

So, if this is a way to go, just let me know and I will remove `MaxHolds` and `MaxFreezes` from everywhere (and introduce **breaking change**), so the `Holds` and `Freezes` will be bounded by composite `RuntimeHoldReason` and `RuntimeFreezeReason` and nobody won't ever need to take care about setting correct constants for `MaxHolds` / `MaxFreezes` without needing integrity test.

Instead, maybe we could add some `try-runtime` check for actual `Holds` / `Freezes` to verify, that actual `RuntimeHoldReason::VARIANT_COUNT` is aligned with stored `Holds` - this could be useful for scenarios, when somebody removes any variant and automatically decrease VARIANT_COUNT, so this `try-runtime` should check it and let us know, that we need to handle removed variant vs on-chain stored data.

Reply:
Yeah, I would prefer to remove `MaxHolds` and `MaxFreezes` as well.

Reply:
Okay, we are on the same track, so I will proceed with removing `MaxHolds` and `MaxFreezes` from everywhere.

Reply:
`MaxHolds` removed from everywhere, and for `MaxFreezes` I created folow-up issue: https://github.com/paritytech/polkadot-sdk/issues/2997 because there is a little bit different situation (please check that issue with more details)